src/utils/SaveFile.js

```diff
@@ -1,79 +1,236 @@
-      fileEntry.createWriter(
-        fileWriter => resolve({ fileWriter, filename: fileEntry.toURL() }),
-        err => reject(err),
-      );
-    });
+const saveZippedSessions = (exportedSessionIds, sessions, installedProtocols) => {
+  if (isElectron()) { // electron save dialog
+    return saveZippedSessionsElectron(exportedSessionIds, sessions, installedProtocols);
+  } else if (isCordova()) {
+    return saveZippedSessionsCordova(exportedSessionIds, sessions, installedProtocols);
+  }
 
-    const writeFile = ({ fileWriter, filename }) => new Promise((resolve, reject) => {
-      const blob = new Blob([data], { type: fileType });
-      fileWriter.seek(0);
-      fileWriter.onwrite = () => resolve(filename); // eslint-disable-line no-param-reassign
-      fileWriter.onerror = err => reject(err); // eslint-disable-line no-param-reassign
-      fileWriter.write(blob);
-    });
+  return Promise.reject('Export not supported on this platform.');
+};
 
-    getFileSystem()
-      .then(getFile)
-      .then(createWriter)
-      .then(writeFile)
-      .then((filename) => {
-        const options = {
-          message: 'Your network canvas file.', // not supported on some apps
-          subject: 'network canvas export',
-          files: [filename],
-          chooserTitle: 'Share file via', // Android only
-          ...shareOptions,
-        };
-        window.plugins.socialsharing.shareWithOptions(options);
-      })
-      .catch(() => { openErrorDialog(); });
-  } else { // browser save to downloads
-    const blob = new Blob([data], { type: fileType });
-    const element = document.createElement('a');
-    element.setAttribute('href', window.URL.createObjectURL(blob));
-    element.setAttribute('download', defaultFileName);
-    element.style.display = 'none';
-    document.getElementById('root').appendChild(element);
-    element.click();
-    document.getElementById('root').removeChild(element);
-    window.URL.revokeObjectURL(blob);
+const saveFile = (exportedSessionIds, sessions, installedProtocols) => {
+  if (exportedSessionIds.length > 1) { // when cvs is possible, it should zip too
+    return saveZippedSessions(exportedSessionIds, sessions, installedProtocols);
   }
+
+  // export single file
+  const sessionId = exportedSessionIds[0];
+  const session = sessions[sessionId];
+  const sessionCodebook = installedProtocols[session.protocolUID].codebook;
+
+  if (isElectron()) {
+    const fs = window.require('fs');
+    const { dialog, shell } = window.require('electron').remote;
+
+    return new Promise((resolve, reject) => {
+      dialog.showSaveDialog({
+        filters: [{ name: 'graphml', extensions: ['graphml'] }],
+        defaultPath: `${session.caseId}_${sessionId}.graphml`,
+      // eslint-disable-next-line consistent-return
+      }, (filename) => {
+        if (filename === undefined) resolve('cancelled');
+        exportElectronFile(fs, session, sessionCodebook, filename)
```

Comment:
Do we need to pass around things like file system objects?:
- They are going to different functions per platform, so if those functions have different requirements, mirroring the interfaces seems a moot point?
- In the case of electron, it looks like `archive()` already provides it's own version of `fs`? Perhaps the same goes for cordova?

That said, I'm supprised to see seperate functions here but for platform requirements to be also handled elsewhere (in `archive()`)? My preference would be for platform checks as close to "the metal" as possible, with identical interfaces so the end result might look like: `exportFile().then(shareResult);`.

Granted that might not be practicable here because of the constraints of Cordova's API?

Reply:
`exportFile()` would be ideal, for sure. I started with what was working in electron on Server, which wouldn't work for Cordova, from the save as dialog to filesystem to archiver. But perhaps more of the middle could be shared? I'll take a look.